gremlin-dotnet/src/Gremlin.Net/Driver/ConnectionPool.cs

```diff
@@ -66,42 +61,35 @@ public async Task<IConnection> GetAvailableConnectionAsync()
 
         private async Task EnsurePoolIsPopulatedAsync()
         {
-            // The pool could have been empty because of connection problems. So, we need to populate it again.
+            // The pool could have been (partially) empty because of connection problems. So, we need to populate it again.
             while (true)
             {
-                var nrOpened = Interlocked.CompareExchange(ref _nrConnections, PoolEmpty, PoolEmpty);
-                if (nrOpened >= _poolSize) break;
-                if (nrOpened != PoolPopulationInProgress)
+                if (NrConnections >= _poolSize) break;
+                var poolState = Interlocked.CompareExchange(ref _poolState, PoolPopulationInProgress, PoolIdle);
+                if (poolState == PoolPopulationInProgress) continue;
+                try
                 {
                     await PopulatePoolAsync().ConfigureAwait(false);
                 }
+                finally
+                {
+                    // We need to remove the PoolPopulationInProgress flag again even if an exception occurred, so we don't block the pool population for ever
+                    Interlocked.CompareExchange(ref _poolState, PoolIdle, PoolPopulationInProgress);
+                }
             }
         }
         
         private async Task PopulatePoolAsync()
         {
-            var nrOpened = Interlocked.CompareExchange(ref _nrConnections, PoolPopulationInProgress, PoolEmpty);
-            if (nrOpened == PoolPopulationInProgress || nrOpened >= _poolSize) return;
-
-            try
-            {
-                var connectionCreationTasks = new List<Task<Connection>>(_poolSize);
-                for (var i = 0; i < _poolSize; i++)
-                {
-                    connectionCreationTasks.Add(CreateNewConnectionAsync());
-                }
-
-                var createdConnections = await Task.WhenAll(connectionCreationTasks).ConfigureAwait(false);
-                foreach (var c in createdConnections)
-                {
-                    _connections.Add(c);
-                }
-            }
-            finally
+            var nrConnectionsToCreate = _poolSize - _connections.Count;
+            var connectionCreationTasks = new List<Task<Connection>>(nrConnectionsToCreate);
+            for (var i = 0; i < nrConnectionsToCreate; i++)
             {
-                // We need to remove the PoolPopulationInProgress flag again even if an exception occurred, so we don't block the pool population for ever
-                Interlocked.CompareExchange(ref _nrConnections, _connections.Count, PoolPopulationInProgress);
+                connectionCreationTasks.Add(CreateNewConnectionAsync());
             }
+
+            var createdConnections = await Task.WhenAll(connectionCreationTasks).ConfigureAwait(false);
```

Comment:
Note that, if there are 2 or more connections being created and one creation fails, the other connection will be open but not returned / or added to the `_connections` list, so it will be leaked.

Reply:
Good catch. Maybe a try-finally where we just `Dispose()` all `createdConnections` in the `finally` block?

I think if a connection establishment fails then we should just let the exception be raised to the user as we can't really handle that.
(Maybe it makes sense later to add a retry here, but we should probably have a logger for cases like that so users can notice such problems.)

Or what do you think?

Reply:
I'm thinking on situations when you end up with a half backed pool, that can be because initially its not possible to create all connections or because one of the connections in the pool closes, after the successfully establishing it.

> Maybe a try-finally where we just Dispose() all createdConnections in the finally block?

Sounds good. I don't have an strong opinion on how to handle it, however you decide is fine by me, as long as we have a consistent behaviour.